platform-sdk/docs/proposals/TSS-Library/TSS-Library.md

```diff
@@ -0,0 +1,808 @@
+  so that signature validation can be done on smart contracts without incurring an excessive gas cost.
+- **Security**: Our produced code should be able to pass internal and external security audits.
+- **Flexibility**: Minimize the impact of introducing support for other elliptic curves.
+- **Independent Release**: When applicable, the new libraries should have the release cycle separate from the platform.They should be implemented in a way that is easy for both platform and block node to depend on.
+
+### Non-Goals
+- Implement support for elliptic curve cryptography in Java.
+- Support any system/architecture other than: Windows amd64, Linux amd64 and arm64, and MacOS amd64 and arm64.
+
+## Changes
+
+### Core Behaviors
+
+The proposed TSS solution is based on Groth21.
+
+Groth21 is a non-interactive, publicly verifiable secret-sharing scheme where a dealer can construct a Shamir secret sharing of a field element
+and confidentially yet verifiably distribute shares to multiple receivers.
+It includes a distributed resharing protocol that preserves the public key but creates a fresh secret sharing of the secret key and hands it to a set of receivers,
+which may or may not overlap with the original set of shareholders.
+non-interactive distributed key generation and key resharing protocols that support Shamir secret sharings of the secret keys.
+
+#### Overview
+This proposal covers the implementation of a tool similar to ssh-keygen to generate those keys, but the generation, persistence, distribution
+and loading of those keys is outside the scope of this proposal.
+
+Participants can hold one or more shares, each of which can be used to sign a message.
+The goal is to generate an aggregate signature which is valid if a threshold number of individual signatures are combined.
+
+Each participant brings their own Elliptic Curve (EC) key pair (private and public). They share their public keys with all other participants while securing their private keys.
+Before the protocol begins, all participants agree on the cryptographic parameters.
+A participant directory is needed when initiating the protocol for the first time.
+This directory includes the number of participants, each participant’s EC public key, and the shares they own.
+
+Each participant generates a random private key and distributes it among the others.
+This allows for verification and aggregation of signatures produced by all shares.
+A value derived from this random private key is created for each share using mathematical functions. Each value is encrypted with the share owner's public key, ensuring only the intended recipient can read it.
+When a message is created, all encrypted values are included. Only the intended receivers can decrypt their respective portions of the secret element.
+This setup allows participants to share secret information securely.
+The message also contains additional information necessary for signature validation.
```

Comment:
Each participant can act as a dealer or receiver.

When generating the keys, they act as dealers, creating the shares for the other participants out of the secret shares they own. They cannot "spy" on other participant's secrets as they don't have access to them. If they are malicious, they could send an invalid message, which other non-malicious participants would detect through the proof and commitments.
When acting as receivers, given that each share is encrypted under each participant's public key, they need access to other participants' private keys to decrypt and "spy" on the shares belonging to other participants.
We can mitigate the risk with a key rotation policy.

@rsinha, please jump in if I need to correct something in the answer.

Reply:
That's right. As long as the adversary has knowledge of fewer than a threshold number of decryption keys, they cannot recover enough information to start forging threshold signatures -- they may learn the shares of the nodes whose keys they have compromised, but that's not enough to recover the full secret.

Reply:
Thank you for that explanation.  I'm still missing something (which suggests the document could use some clarifying text).

What's still not clear to me, is how the nodes are prevented from obtaining a threshold number of unencrypted shares.
Perhaps it's just a missing step or word confusion; I'll list what I think is happening, based on the proposal description, below.
1. Nodes gossip random private keys.
   - "Each participant generates a random private key and distributes it among the others."
1. A dealer (one, many?) receives all the random values, generates a share for each random value
   - "A value derived from this random private key is created for each share using mathematical functions."
1. The dealer, after deriving the value, encrypts the share
   - "Each value is encrypted with the share owner's public key, ensuring only the intended recipient can read it."
      - Couldn't a malicious actor inspect all of the values here, and possibly encrypt chosen share values?
      - Secondary, is every node doing this, in concert, or a single dealer?
         - I am assuming every node completes this step.
1. Each node, having calculated _all_ of the encrypted shares, gossips out the encrypted shares.
   - "When a message is created, all encrypted values are included."

This next bit is where it gets confusing.
 - "Upon receiving a threshold number of messages, each participant decrypts the information encrypted with their public key, aggregates it, and generates a private key for each owned share."

Would that mean each node is computing a _partial_ share for each other node?  That might be the part I am missing, but I'm not certain.

Reply:
Let me work on the rewording. I see the confusion.

Reply:
is this more clear:
```
Each participant generates portions of a secret share and distributes them among the other participants using the following process:

1. A random private key is created and mathematically split into a known number of total shares.
2. Each portion is encrypted with the share owner's public key, ensuring only the intended recipient can read it.
3. A message is created that includes all encrypted values so that only the intended recipients can decrypt their respective portions of the secret share.

This setup allows participants to share secret information securely. The message also contains additional information necessary for its validation.

Upon receiving a threshold number of messages, each participant:

1. Validates the message and decrypts the information encrypted with their public key.
2. Aggregates the decrypted information to generate a private key for each owned share.
3. Retrieves a public key for each share in the system to validate signatures.

Individual signing can then begin. Participants use the private information of their shares to sign messages.

An aggregate signature is created when signatures from at least the threshold number of parties are combined. This aggregate signature can be validated using the combined value of the public shares in the directory.

The process restarts whenever the number of participants or the shares assigned to each change. However, the initially generated group public key remains unchanged to maintain consistency. New secret information for shares is created using existing data, ensuring the aggregate signature can still be verified with the original group public key.
```

Reply:
Much clearer.  We might add a sub-point to express how these independent actions are effectively coordinated.  Perhaps something like the following (I am somewhat guessing that this is the case; please correct if I have it wrong):
> 1. A random private key is created and mathematically split into a known number of total shares.
>>    * These shares are related to shares from other participants through a pre-arranged polynomial commitment.